# Co-relation analysis

In [1]:
import pandas as pd
dataset = pd.read_csv('C:/Users/kulka/OneDrive/Desktop/me/ML-HHH/lets upgrade/tutorials/general_data.csv')

In [2]:
dataset.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [3]:
dataset.tail()

Age Attrition BusinessTravel              Department  DistanceFromHome  \
4405   42        No  Travel_Rarely  Research & Development                 5   
4406   29        No  Travel_Rarely  Research & Development                 2   
4407   25        No  Travel_Rarely  Research & Development                25   
4408   42        No  Travel_Rarely                   Sales                18   
4409   40        No  Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
4405          4        Medical              1        4406  Female  ...   
4406          4        Medical              1        4407    Male  ...   
4407          2  Life Sciences              1        4408    Male  ...   
4408          2        Medical              1        4409    Male  ...   
4409          3        Medical              1        4410    Male  ...   

      NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
4405                 3.0      Y                17              8   
4406                 2.0      Y                15              8   
4407                 0.0      Y                20              8   
4408                 0.0      Y                14              8   
4409                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
4405               3                        0                     2  
4406               3                        0                     2  
4407               4                        1                     2  
4408               9                        7                     8  
4409              21                        3                     9  

[5 rows x 24 columns]

In [4]:
dataset.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [5]:
dataset1 = dataset.drop_duplicates()
dataset2 = dataset.dropna()

## checking correlation between variables

In [6]:
from sklearn.preprocessing import LabelEncoder

In [7]:
le = LabelEncoder()
dataset2['Attrition'] = le.fit_transform(dataset2['Attrition'])
dataset2.head()

<ipython-input-7-8ae24e6ccbc4>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset2['Attrition'] = le.fit_transform(dataset2['Attrition'])


Age  Attrition     BusinessTravel              Department  \
0   51          0      Travel_Rarely                   Sales   
1   31          1  Travel_Frequently  Research & Development   
2   32          0  Travel_Frequently  Research & Development   
3   38          0         Non-Travel  Research & Development   
4   32          0      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                 6          2  Life Sciences              1           1   
1                10          1  Life Sciences              1           2   
2                17          4          Other              1           3   
3                 2          5  Life Sciences              1           4   
4                10          1        Medical              1           5   

   Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0  Female  ...                 1.0      Y                11              8   
1  Female  ...                 0.0      Y                23              8   
2    Male  ...                 1.0      Y                15              8   
3    Male  ...                 3.0      Y                11              8   
4    Male  ...                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

### Correrlation matrix

In [8]:
from scipy.stats import pearsonr
from sklearn.preprocessing import LabelEncoder

In [9]:
dataset2.corr()

Age  Attrition  DistanceFromHome  Education  \
Age                      1.000000  -0.158399          0.007376  -0.033900   
Attrition               -0.158399   1.000000         -0.009449  -0.017106   
DistanceFromHome         0.007376  -0.009449          1.000000  -0.007491   
Education               -0.033900  -0.017106         -0.007491   1.000000   
EmployeeCount                 NaN        NaN               NaN        NaN   
EmployeeID               0.008105  -0.004621         -0.000326  -0.009389   
JobLevel                -0.001137  -0.012382         -0.039990   0.045822   
MonthlyIncome           -0.045163  -0.030160         -0.022757   0.007289   
NumCompaniesWorked       0.299527   0.042831         -0.014449  -0.016210   
PercentSalaryHike       -0.032561   0.033153          0.037720  -0.041054   
StandardHours                 NaN        NaN               NaN        NaN   
StockOptionLevel        -0.031504  -0.008164          0.009353   0.002386   
TotalWorkingYears        0.680037  -0.169670          0.009574  -0.009228   
TrainingTimesLastYear   -0.028962  -0.047586         -0.008957   0.009939   
YearsAtCompany           0.311281  -0.133003          0.030746   0.005997   
YearsSinceLastPromotion  0.215650  -0.031423          0.002243   0.023457   
YearsWithCurrManager     0.201580  -0.154692          0.021773   0.005645   

                         EmployeeCount  EmployeeID  JobLevel  MonthlyIncome  \
Age                                NaN    0.008105 -0.001137      -0.045163   
Attrition                          NaN   -0.004621 -0.012382      -0.030160   
DistanceFromHome                   NaN   -0.000326 -0.039990      -0.022757   
Education                          NaN   -0.009389  0.045822       0.007289   
EmployeeCount                      NaN         NaN       NaN            NaN   
EmployeeID                         NaN    1.000000 -0.003090       0.007865   
JobLevel                           NaN   -0.003090  1.000000       0.046688   
MonthlyIncome                      NaN    0.007865  0.046688       1.000000   
NumCompaniesWorked                 NaN    0.000719 -0.009759      -0.021446   
PercentSalaryHike                  NaN   -0.004877  0.010874       0.004607   
StandardHours                      NaN         NaN       NaN            NaN   
StockOptionLevel                   NaN   -0.013488  0.000365       0.027242   
TotalWorkingYears                  NaN   -0.001688 -0.036293      -0.034398   
TrainingTimesLastYear              NaN   -0.012102 -0.031931       0.049780   
YearsAtCompany                     NaN    0.004117 -0.063360       0.000088   
YearsSinceLastPromotion            NaN    0.000814 -0.059680       0.064470   
YearsWithCurrManager               NaN    0.009079 -0.053898       0.023095   

                         NumCompaniesWorked  PercentSalaryHike  StandardHours  \
Age                                0.299527          -0.032561            NaN   
Attrition                          0.042831           0.033153            NaN   
DistanceFromHome                  -0.014449           0.037720            NaN   
Education                         -0.016210          -0.041054            NaN   
EmployeeCount                           NaN                NaN            NaN   
EmployeeID                         0.000719          -0.004877            NaN   
JobLevel                          -0.009759           0.010874            NaN   
MonthlyIncome                     -0.021446           0.004607            NaN   
NumCompaniesWorked                 1.000000           0.030064            NaN   
PercentSalaryHike                  0.030064           1.000000            NaN   
StandardHours                           NaN                NaN            NaN   
StockOptionLevel                   0.016291           0.012104            NaN   
TotalWorkingYears                  0.238807          -0.018480            NaN   
TrainingTimesLastYear             -0.031335          -0.036720            NaN   
Years

The above is the entire correlation table. Since, we want to look at correlation with attrition only as it will help us determine the solution of our problem statement. Hence, in the next part, we will check the influence of each attribute on attrition. We will perfom the analysis on the attributes: 'Age, DistanceFromHome, Gender, MonthlyIncome, NumCompaniesWorked, PercentSalaryHike, TotalWorkingYears, YearsAtCompany, YearsSinceLastPromotion, YearsWithCurrManager'

In [10]:
dataset2.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

### Attrition Vs age

In [11]:
stats,p = pearsonr(dataset2.Attrition,dataset2.Age)
print(stats,p)

-0.15839867954096684 5.1265982193804103e-26


As we can see, the attribute 'Age' has a negative correlation coefficient with attrition. As age increases, the chances of attrition slightly decrease. The p value also tells that, the null hypothesis is rejected and there is a correlation between age and attrition

### Attrition Vs DistanceFromHome

In [12]:
stats,p = pearsonr(dataset2.Attrition,dataset2.DistanceFromHome)
print(stats,p)

-0.009448638515156229 0.5317715668019558


The r value = -0.009 suggests that there is a very low negative correlation between attrition and DistanceFromHome. The p-value > 0.05 confirms the same. Hence, we conclude that DistanceFromHome doesn't significantly contribute to correlation

### Attrition Vs Gender

In [13]:
dataset2['Gender'] = le.fit_transform(dataset2['Gender'])
dataset2.head()

<ipython-input-13-2f44f593415a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset2['Gender'] = le.fit_transform(dataset2['Gender'])


Age  Attrition     BusinessTravel              Department  \
0   51          0      Travel_Rarely                   Sales   
1   31          1  Travel_Frequently  Research & Development   
2   32          0  Travel_Frequently  Research & Development   
3   38          0         Non-Travel  Research & Development   
4   32          0      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                 6          2  Life Sciences              1           1   
1                10          1  Life Sciences              1           2   
2                17          4          Other              1           3   
3                 2          5  Life Sciences              1           4   
4                10          1        Medical              1           5   

   Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0       0  ...                 1.0      Y                11              8   
1       0  ...                 0.0      Y                23              8   
2       1  ...                 1.0      Y                15              8   
3       1  ...                 3.0      Y                11              8   
4       1  ...                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [14]:
stats,p = pearsonr(dataset2.Attrition,dataset2.Gender)
print(stats,p)

0.018396425194440603 0.2233996608682602


r value = 0.018 indicates low positive correlation. The p-value > 0.05 suggests not a significant impact. Thus, there is no significant impace of gender on attrition

### Attrition Vs monthly income

In [15]:
stats,p = pearsonr(dataset2.Attrition,dataset2.MonthlyIncome)
print(stats,p)

-0.030160293808460616 0.045890862744719166


r = -0.03 indicates moderate negative correlation. p value < 0.05 is just below 0.05. Thus, we can say, there is moderately significant impact of monthly income on attrition. As monthly income increases, attrition drops mildly.

### Attrition Vs Num of companies worked

In [16]:
stats,p = pearsonr(dataset2.Attrition,dataset2.NumCompaniesWorked)
print(stats,p)

0.042830567244720494 0.00457205712162295


r = 0.042 is a moderately positive correlation. p value < 0.05. Thus, we can say, there is a significant impact of number of companies worked and attrition. More the companies worked, more is chance of attrition

### Attrition Vs Percent Salary Hike

In [17]:
stats,p = pearsonr(dataset2.Attrition,dataset2.PercentSalaryHike)
print(stats,p)

0.0331530371354668 0.028192446935093166


r = 0.033 is a moderately positive correlation. p-value < 0.05 just slightly. Thus, Percent salary hike just affects attrition i.e. more is the percent salary hike, more is chance of attrition (which should not have been true)

### Attrition Vs Total Working Years

In [18]:
stats,p = pearsonr(dataset2.Attrition,dataset2.TotalWorkingYears)
print(stats,p)

-0.16966991684723928 1.1645434967089207e-29


r = -0.17 is a strong negative correlation. p value < 0.05. Thus, there is a significant impact of Total working years on attrition. More are the total working years, less is the attrition

### Attrition Vs  YearsAtCompany

In [19]:
stats,p = pearsonr(dataset2.Attrition,dataset2.YearsAtCompany)
print(stats,p)

-0.13300261842521555 9.476118084838662e-19


r = -0.13 is a strong negative correlation. p value < 0.05. Thus, there is a significant impact of years at company on attrition. More are the years at company, less is the attrition chance.

### Attrition Vs YearsSinceLastPromotion

In [20]:
stats,p = pearsonr(dataset2.Attrition,dataset2.YearsSinceLastPromotion)
print(stats,p)

-0.03142315056331017 0.03752293607392567


r = -0.03 is a moderately negative correlation. p < 0.05 just. Thus, there is a moderately significant impact of Years Since Last Promotion on attrition. As years since last promotion increase, attrition chances slightly decrease.

### Attrition Vs YearsWithCurrManager

In [21]:
stats,p = pearsonr(dataset2.Attrition,dataset2.YearsWithCurrManager)
print(stats,p)

-0.15469153690287202 7.105369646772844e-25


r = -0.15 is a strong negative correlation. p < 0.05. Thus, there is a strong correlation between attrition and years with current manager. As years with current manager increase, theree attrition chances decrease.